src/context/values.js

```diff
@@ -0,0 +1,720 @@
+   * @private
+   */
+  maybeUpdated_(used, value) {
+    const {prop, value: oldValue} = used;
+    const {key} = prop;
+    const usedByKey = this.usedByKey_;
+    if (
+      oldValue === value ||
+      used !== (usedByKey && usedByKey.get(key)) ||
+      !this.isConnected_()
+    ) {
+      // Either the value didn't change, or no one needs this value anymore.
+      return;
+    }
+
+    used.value = value;
+
+    // Notify subscribers.
+    const {subscribers} = used;
+    subscribers.forEach((handler) => {
+      handler(value);
+    });
+  }
+
+  /**
+   * The used value calculation algorithm.
+   *
+   * @param {!UsedDef} used
+   * @param {boolean} refreshParent
+   * @return {*|undefined} The used value.
+   * @private
+   */
+  calc_(used, refreshParent) {
+    devAssert(this.isConnected_());
+
+    const {prop, depValues} = used;
+    const {key, compute, defaultValue} = prop;
+
+    const inputsByKey = this.inputsByKey_;
+    const inputs = inputsByKey && inputsByKey.get(key);
+    const inputValues = inputs && inputs.values;
+
+    // Calculate parent value.
+    const needsParent = calcNeedsParent(prop, inputValues);
+
+    // Refresh parent if requested.
+    if (refreshParent || needsParent != Boolean(used.parentValueNode)) {
+      const newParentValueNode = needsParent
+        ? findParent(this.contextNode_, hasInput, prop, /* includeSelf */ false)
+        : null;
+      this.updateParentValueNode_(used, newParentValueNode);
+    }
+
+    // If no parent node is found, use the default value.
+    const parentValue = defined(used.parentValue)
+      ? used.parentValue
+      : needsParent && !used.parentValueNode && inputValues
+      ? defaultValue
+      : undefined;
+
+    // Calculate the "used" value.
+    let newValue = undefined;
+    const ready =
+      depValues.every(defined) && (!needsParent || defined(parentValue));
+    if (ready) {
+      const {node} = this.contextNode_;
+      if (inputValues && !compute) {
+        newValue = inputValues[0];
+      } else if (isResursive(prop)) {
+        if (inputValues) {
+          // The node specifies its own input values and they need to be
+          // recomputed with parent and dep values.
+          newValue = compute(
+            node,
+            inputValues,
+            parentValue,
+            // This is `...depValues`, but faster. See MAX_DEPS.
+            depValues[0],
+            depValues[1],
+            depValues[2]
+          );
+        } else if (defined(parentValue)) {
+          // The node doesn't specify its own value, but parent is available.
+          // Since parent is available, it means that the node is recursive.
+          newValue = parentValue;
+        }
+      } else if (compute) {
+        newValue = compute(
+          node,
+          inputValues || EMPTY_ARRAY,
+          // This is `...depValues`, but faster. See MAX_DEPS.
+          depValues[0],
+          depValues[1],
+          depValues[2]
+        );
+      }
```

Comment:
Is it possible to have a `ready` value that does not follow one of these three cases?

Reply:
I'm not super happy with that `if` as well. But basically it ends with the `} else { newValue = undefined }`, which is already set above. So it was kind of weird to put it there as well. I tried to re-express this if but it always has a few cases that fall into the `newValue = undefined` branch like this.